投稿削除機能の実装(ユーザー側)

Rihitonnnu opened the pull request:
## 今回のPRで行っていること
- 投稿削除機能の実装

## UI変更点
### 変更
投稿詳細画面：削除ボタン(ゴミ箱アイコン)を追加しました
<img width="949" alt="image" src="https://user-images.githubusercontent.com/69156872/201288161-58a7640f-b832-4516-b425-c5385ed71f17.png">

Rihitonnnu committed:
destroyメソッド追加

Rihitonnnu committed:
投稿削除のレスポンスのテストケース追加

Rihitonnnu committed:
投稿データを削除するdestroyArticleメソッド追加

Rihitonnnu committed:
投稿データ削除処理のテストケース追加

Rihitonnnu committed:
削除ボタン(ゴミ箱アイコン)のコンポーネント作成

Rihitonnnu committed:
投稿詳細画面に削除ボタンを追加

YutakaOkabe commented on resources/views/user/article/show.blade.php:
```diff
@@ -14,10 +14,19 @@
-                                        <h1 class="font-bold text-3xl text-black">{{ $article->title }}</h1>
+                                        <div class="flex">
+                                            <h1 class="font-bold text-3xl text-black">{{ $article->title }}</h1>
+                                            <form onsubmit="return confirm('投稿を削除してもよろしいですか？')" action="{{ route('user.article.destroy',['article'=>$article->id]) }}" method="post" >
```
詳細画面でタイトルの横に削除ボタンとは中々挑戦的なUIだねw 好みなので全然良いけどw

YutakaOkabe commented on tests/Feature/User/ArticleControllerTest.php:
```diff
@@ -165,4 +165,24 @@ public function ログインしていない状態で投稿内容を更新した
+    {
+        $this->actingAs($this->user, 'users');
+
+        $response = $this->delete(route('user.article.destroy', ['article' => Article::factory()->create()->id]));
```
今思ったけど、他でログインしてるユーザーが自分の記事の詳細画面とか削除URL叩いたら、閲覧も削除もできちゃう気がするw
どうしたらいいか考えといてー！（丸投げ

YutakaOkabe reviewed (changes requested):
ごめん今日中に見れなかった！続きは来週みる！🙇‍♂️
他のユーザーに自分の記事削除させないようにする方法考えといてー！

Rihitonnnu committed:
処理を行う投稿が自分の投稿であるかどうかを判定し違う場合にはアクセス制限をするように変更

Rihitonnnu committed:
アクセス制限の追加によるテスト内容の一部変更

$this->user->idと投稿のuser_idが同じでない場合403へリダイレクトするので記事のテストデータのuser_idを$this->user->idとおなじになるように変更しました

Rihitonnnu committed:
アクセス制限を行うArticlePolicyファイルの作成

Rihitonnnu committed:
ArticlePolicyの登録

Rihitonnnu committed:
削除画面を詳細画面から一覧画面の欄に移動

Rihitonnnu committed:
phpdoc書き忘れの修正

Rihitonnnu commented:
@YutakaOkabe 
見て頂いた部分の修正をしたのでレビューお願いします。
また以下がUIで変更した部分（削除ボタンの移動）のスクショです

投稿詳細画面：削除ボタンを消しました。
<img width="960" alt="image" src="https://user-images.githubusercontent.com/69156872/201565344-8ae1ea7e-2bbf-402e-953a-3b233f18e3b4.png">


投稿一覧画面：削除ボタンを追加しました。
<img width="959" alt="image" src="https://user-images.githubusercontent.com/69156872/201565496-366b8b3b-a7cc-47a5-9a21-22ba24012e72.png">

YutakaOkabe commented on app/Http/Controllers/User/ArticleController.php:
```diff
@@ -56,7 +56,11 @@ public function store(CreateRequest $request)
-        return view('user.article.show', ['article' => Article::find($id),]);
+        $article = Article::find($id);
+
+        $this->authorize('show', $article); //取得した投稿がユーザーが表示可能かどうか判定
```
積極的にLaravelの機能使ってていいね 👍
ここはリソースコントローラー使ってるんだったら `authorizeResource` 使ったほうが綺麗かもなー

https://readouble.com/laravel/9.x/ja/authorization.html#:~:text=%E4%BD%9C%E6%88%90%E5%8F%AF%E8%83%BD%0A%7D-,%E3%83%AA%E3%82%BD%E3%83%BC%E3%82%B9%E3%82%B3%E3%83%B3%E3%83%88%E3%83%AD%E3%83%BC%E3%83%A9%E3%81%AE%E8%AA%8D%E5%8F%AF,-%E3%83%AA%E3%82%BD%E3%83%BC%E3%82%B9%E3%82%B3%E3%83%B3%E3%83%88%E3%83%AD%E3%83%BC%E3%83%A9

Rihitonnnu replied:
ありがとうございます！最初はauthorizeResourceで実装していたのですが自分の投稿なのに403エラーが起こってしまいうまくいかなかったです。調べてみると以下のサイトでコントローラのメソッドの引数がデフォルトと異なるとうまくいかない(?)ということがあるらしいということが書いてありました。
https://biz.addisteria.com/polilcy-resource-controller/#toc7

すべてのメソッドに対して認可を行いたいわけではなかったので、認可したいメソッドに対して一つずつ記述していたのですがやっぱりきれいじゃないですよね...修正してみます！

YutakaOkabe replied:
ほーなるほど、そういう落とし穴もあるのね 😅
ごめんそこまでは知らなかったので、ちょっと調べてみて解決できなそう & 使い勝手悪ければいまのままでよいよ！

Rihitonnnu replied:
とりあえずうまくいく方法がないか調べてみます！無理そうだったら先に進みます🙇‍♀

YutakaOkabe commented on resources/views/user/article/show.blade.php:
```diff
@@ -14,10 +14,12 @@
                                 <div class="flex flex-col sm:flex-row mt-10">
                                     <div class="w-full sm:pl-8 border-gray-200 sm:border-t-0 border-t mt-4 pt-4 sm:mt-0 text-center sm:text-left">
-                                        <h1 class="font-bold text-3xl text-black">{{ $article->title }}</h1>
+                                        <div class="flex">
```
もうflexいらなそうー

Rihitonnnu replied:
了解です！修正しておきます！

YutakaOkabe reviewed (approved):
テキトーにコメントしちゃってたのにちゃんと調べて実装してるのえらい！！
他にも気になったとこほとんどなかった！ちょっとコメント入れたところ以外はLGTM〜！

Rihitonnnu committed:
authorizeResourceの導入に伴いそれぞれのメソッドの引数を変更

Rihitonnnu committed:
authorizeResourceの導入に伴いPolicyのメソッドを変更

Rihitonnnu committed:
authorizeResourceの導入に伴い投稿の削除処理の引数と記述変更

Rihitonnnu committed:
authorizeResourceの導入に伴い投稿の削除処理のテストケースの記述変更

Rihitonnnu committed:
不要なflexの削除

Rihitonnnu commented:
この記事を参考にArticleControllerのメソッドの引数を変更した所、authorizeResourceでの実装がうまくいきました！
https://tohokuaiki.hateblo.jp/entry/2020/07/31/authorizeResource%E3%81%8C%E5%8A%B9%E3%81%8B%E3%81%AA%E3%81%84%E3%81%98%E3%82%83%E3%82%93%E3%81%A3%E3%81%A6%E6%82%A9%E3%82%93%E3%81%A7%E3%81%9F%E3%82%89%E2%80%A6

Rihitonnnu commented:
レビューありがとうございました！
